### PR TITLE
GH-2458 sh:hasValue and dash:hasValueIn support as node shapes

### PR DESCRIPTION
**Pull request by hmottestad**
GitHub issue resolved: #2458  

Briefly describe the changes proposed in this PR:

Make sh:hasValue and dash:hasValueIn work when part of a sh:NodeShape. This fixes when they are used as part of a logical operator like sh:or. 

---- 
PR Author Checklist (see the [contributor guidelines](https://github.com/eclipse/rdf4j/blob/master/.github/CONTRIBUTING.md) for more details):

 - [ ] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [ ] I've added tests for the changes I made
 - [ ] I've applied [code formatting](https://github.com/eclipse/rdf4j/blob/master/.github/CONTRIBUTING.md#code-formatting) (you can use `mvn process-resources` to format from the command line)
 - [ ] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change
 - [ ] every commit has been [signed off](https://stackoverflow.com/questions/1962094/what-is-the-sign-off-feature-in-git-for)

Note: we merge all feature pull requests using [squash and merge](https://help.github.com/en/github/administering-a-repository/about-merge-methods-on-github#squashing-your-merge-commits). See [RDF4J git merge strategy](https://rdf4j.org/documentation/developer/merge-strategy/) for more details.

### CODE REVIEW

**Comment by hmottestad** [comment]
@jeenbroekstra I would appreciate it if you could take a look at this. I would like to merge it and release a bug fix release once ontotext gives it a quick test.

**Comment by jeenbroekstra** [comment]
I'll take a look. While I do that, can you please link to the issue and update the PR description?

File: core/sail/shacl/src/test/java/org/eclipse/rdf4j/sail/shacl/TempTest.java [hunk 1]
```diff
@@ -500,4 +500,50 @@ private void add(SailRepositoryConnection connection, String data) throws IOExce
 		connection.commit();
 	}
 
+	@Test()
+	public void temp() throws Throwable {
+
```
Comment (jeenbroekstra) [inline core/sail/shacl/src/test/java/org/eclipse/rdf4j/sail/shacl/TempTest.java:505]: Please remove this class.

File: core/sail/shacl/src/test/resources/test-cases/hasValueIn/or/shacl.ttl [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+@prefix sh: <http://www.w3.org/ns/shacl#> .
+@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
+@prefix dash: <http://datashapes.org/dash#>.
+
+ex:shape1
+	a sh:NodeShape ;
+	sh:targetClass ex:Person ;
+	sh:property [
+		sh:path ex:info ;
+		sh:or ([ dash:hasValueIn ("blue") ; ] [ dash:hasValueIn ("green") ; ] [ dash:hasValueIn ("red"); ]) ;
```
Comment (jeenbroekstra) [inline core/sail/shacl/src/test/resources/test-cases/hasValueIn/or/shacl.ttl:15]: I don't object to having a test for a corner case, but isn't this an odd use of `dash:hasValueIn`? I thought the point was to allow use of a list so you don't have to use an or constraint at all. In other words, isn't the above equivalent to just saying:

     sh:path ex:info;
     dash:hasValueIn ("blue" "green" "red");

File: core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/AST/AndPropertyShape.java [hunk 3]
```diff
@@ -227,15 +227,45 @@ public PlanNode getAllTargetsPlan(ConnectionsGroup connectionsGroup, boolean neg
...
+					.map(Optional::get)
+					.collect(Collectors.joining(" ) && ( ", "( ",
+							" )"));
+
+			String query = pathQuery1 + "\n FILTER (! EXISTS {\n" + pathQuery1.replaceAll("(?m)^", "\t")
+					+ "\n\tFILTER(!(" + collect + "))\n})";
+
+			String pathQuery2 = getPath().getQuery(targetVar, randomVariable(), null);
+
+			query = "{\n#VALUES_INJECTION_POINT#\n " + query.replaceAll("(?m)^", "\t")
```
Comment (jeenbroekstra) [inline core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/AST/AndPropertyShape.java:250]: what is `VALUES_INJECTION_POINT`?

**Review by jeenbroekstra — CHANGES_REQUESTED** [review]
Like I mentioned on the actual issue ticket: I don't follow what exact problem you're solving. Can you clarify  on the issue and by updating the PR description?

File: core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/AST/AndPropertyShape.java (hunk 3 above)
Comment (hmottestad) [reply core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/AST/AndPropertyShape.java:250]: It's where we inject SPARQL values. Since we have unions we end up introducing a new scope, which leads to poor performance. So we add points where we can inject sparql values, for the cases where we need to run the query against a subset of nodes.

Comment (jeenbroekstra) [reply core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/AST/AndPropertyShape.java:250]: Ah I see. ObNitpick: it might be good to declare a constant for this substitution pattern.

File: core/sail/shacl/src/test/resources/test-cases/hasValueIn/or/shacl.ttl (hunk 2 above)
Comment (hmottestad) [reply core/sail/shacl/src/test/resources/test-cases/hasValueIn/or/shacl.ttl:15]: That would be equivalent to:
```
sh:or ([sh:path ex:info ; dash:hasValueIn ("blue") ; ] [sh:path ex:info ; dash:hasValueIn ("green") ; ] [sh:path ex:info ; dash:hasValueIn ("red"); ]) ;
```

Comment (jeenbroekstra) [reply core/sail/shacl/src/test/resources/test-cases/hasValueIn/or/shacl.ttl:15]: Ah I see! I think I'm beginning to understand what the actual issue here is. At least, I think so...

File: core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/AST/AndPropertyShape.java [hunk 4]
```diff
@@ -227,15 +227,45 @@ public PlanNode getAllTargetsPlan(ConnectionsGroup connectionsGroup, boolean neg
...
-						.map(propertyShape -> propertyShape.buildSparqlValidNodes(targetVar))
-						.reduce((a, b) -> a + "\n" + b))
-				.filter(Optional::isPresent)
-				.map(Optional::get)
-				.reduce((a, b) -> a + "\n" + b)
-				.orElse("");
+
+		if (hasOwnPath()) {
+			// within property shape
+			String objectVariable = "?b";
```
[the remark this answers is not in the document]
Comment (hmottestad) [reply core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/AST/AndPropertyShape.java:233]: I'll take a look and check.

**Comment by hmottestad** [comment]
I believe this should be done now. I found another edge case to handle, but that's fixed now.

**Review by jeenbroekstra — APPROVED** [review]
LGTM, thanks!

**Comment by jeenbroekstra** [comment]
@hmottestad on the issue ticket @rdstn pointed out that the whole thing is counter-intuitive and we perhaps should add some documentation to explain how to use this. Is that something you want to tackle as part of this issue, or will you log a followup for that (or do you think it's not our problem to document this)?

**Comment by hmottestad** [comment]
The issue is with SHACL being counter intuitive and not so much this implementation. So I'm not sure where to document it.
